feat(templating-router): async property in route-href which prevents …

michie1 opened the pull request:
…errors for (yet) undefined params

Issue #21 noted error "A value is required for route parameters ... in route ..."
when the params data is available after an asynchronous call.
This feature adds the async property. If enabled, the bind method does not process the url
when the params include undefined values.

EisenbergEffect commented:
@michie1 Before we can proceed further we would need a CLA signature.

michie1 commented:
@EisenbergEffect I tried several times to sign the CLA. The page responses I signed, but a day later the CLA check is not completed yet.
![cla-signed](https://cloud.githubusercontent.com/assets/1347371/19529410/09ec2a68-9630-11e6-8279-0305246e8222.png)

EisenbergEffect commented:
You probably  have different email addressed associated with your PR than with your account.

CasiOo commented:
Would be nice with a comment explaining what the code is doing. It isn't obvious to me at least.

michie1 commented:
@EisenbergEffect That was indeed the issue. Fixed!

@CasiOo 
> // with the async attribute set to true, undefined values of route parameters won't yield requirement errors.
Would this clarify the purpose of the code for you?

CasiOo commented:
Yes I now understand what the code is doing, and why it is there.

I'm not sure why the check for an undefined value is needed. What if undefined is expected when using optional route parameters?
In async routes the params object starts by either being undefined or null, right? Why should it even have any properties to begin with?

What happens if the `route` is changed (not the params) before the async params result?

What about making params bindable accept a promise value? It could then simply wait for the promise to either resolve or reject.


Set `async` to true, and you'll have to update the params

michie1 commented:
@CasiOo 
> I'm not sure why the check for an undefined value is needed.

An undefined value will result in an error in processChange(). (ERROR [route-href] Error: A value is required for route parameter ...) To prevent the error, we don't call processChange if there is any undefined value.

> What if undefined is expected when using optional route parameters?

With the async feature the first error is prevented in bind(). 
After the async callback, attributeChanged will probably cause another error. 

> In async routes the params object starts by either being undefined or null, right?

Yes, undefined.

> Why should it even have any properties to begin with?

What do you mean?

> What happens if the route is changed (not the params) before the async params result?

How do you change the route? You mean changines <this> in ``route-href="route: <this>"``?

> What about making params bindable accept a promise value? It could then simply wait for the promise to either resolve or reject.

Could you describe the steps to implement this?

CasiOo commented:
> An undefined value will result in an error in processChange(). (ERROR [route-href] Error: A value is required for route parameter ...) To prevent the error, we don't call processChange if there is any undefined value.
If the async params starts out as either undefined or null there would be no need for checking any values for undefined. A call to `processChange()` should always be prevented in `bind` if async, yes?

> How do you change the route? You mean changines in route-href="route: <this>"?
If either `route` or `attribute` changes it will invoke `processChange()`. If our async params haven't yet been updated at this point, it will result in the error we're trying to avoid.
```
<a route-href="route.bind: route; params.bind: params; async: true">

activate() {
    // An error will be raised here because route is updated before our async params
    this.route = 'testModule';
    this.params = {};
}
```

I'm not even sure what should happen if you change the `async` attribute at a later point after bind. Is it just a 'dead' bindable after the first bind?

michie1 commented:
> If the async params starts out as either undefined or null there would be no need for checking any values for undefined. A call to processChange() should always be prevented in bind if async, yes?

Good question! The checking for the undefined values is important, because it is possible the async operation is already finished before the bind method is called. This won't trigger attributeChanged. 

I haven't considered the case when routes/attributes are dynamically changed. 

> async: true only indicates that the params can be async? Is it never intended to have any of the other bindables being async changed?

Yes. Yes. The bindable name asyncParam could be a better alternative, because it's more specific.

EisenbergEffect commented:
Where are we at with this?

michie1 commented:
@EisenbergEffect 
1. I just changed the name of the attribute to async-param and added a comment to the code.
2. I'm waiting for help how to write a test for this. 
3. @CasiOo had an idea how to solve this problem in a neater way with promises and I like to hear the steps how to implement this.
4. @kristianmandrup likes the solution.
5. @JoshMcCullough Can you check if this pull request solves your problem?

kristianmandrup commented:
Good to see there is finally some movement on this... 
I have sadly since given up on Aurelia and shifted to Vue 2, which combines the best of Angular, Aurelia and React etc. in a very clean and flexible fashion. Wish you all the best.

EisenbergEffect commented:
@CasiOo Would love to hear you chime in with your ideas.

CasiOo commented:
If we were to follow how async programming is usually dealt with, then we'll have to wait for the async operation to complete. The suggested solution doesn't wait for an async operation to complete, instead it waits for data to change. I would, as a developer, feel more in control if I get to choose just the right moment when the route-href should update. I'm not convinced it is a solid solution to wait for the first instance change of `param`.

> I haven't considered the case when routes/attributes are dynamically changed.
We'll need to have this case fixed before a release.

Promises are no stranger to the router, and are the preferred way of dealing with async operations in modern JavaScript. To me it makes sense to somehow utilize promises in solving the issue.

1) The quick and dirty way of implementing a fix, is letting the `params` bindable understand promises. If it sees a promise, then it awaits the result. I doubt this will be an acceptable solution for typescript users, since you'll then have a bindable accepting type <any>  and <Promise> where promise has a special meaning. I'm not experienced in API design, but in any type safe language it would probably have been a no no.

2) We tell the route-href when to update. Developers are known with the term `async`, so lets use it.
Here we wait for the `asyncUserPromise` promise to resolve, and below we wait for the given signal.
```
<a route-href="route: 'user'; params.bind: { id: user.id }; await: asyncUserPromise"></a>
<a route-href="route: 'user'; params.bind: { id: user.id }; await-signal: 'user-data-ready-signal'"></a>
```

3) Support active/inactive route-href's through a bindable? `active.bind: activateUserAnchors`

4) Implement a `waitFor` binding behavior and attach it to the `params` bindable.
```<a route-href="route: 'user'; params.bind: { id: user.id } & waitFor: asyncUserPromise"></a>```
I'm not sure such a binding behavior could work in our case. It's difficult to see how the binding behavior can tell the route-href not to update.

Personally I'm in favor of my second suggestion.

michie1 commented:
@CasiOo Thank you for your suggestions. 

I decided I'm not going to (try) to implement on of the suggestions. Is there anyone else interested? 

@EisenbergEffect You can decide if you want to accept or reject my pull request.

CasiOo commented:
I'll happily implement the suggestion I came up with. Though, I don't want to waste my time implementing the feature if the Aurelia team doesn't think it's a good idea.

@michie1 It would indeed be great if the Aurelia team could tell which direction to go.

EisenbergEffect commented:
I'll review again shortly.

EisenbergEffect commented:
@bryanrsmith Do you have any thoughts on this?

CasiOo commented:
I guess this issue is having low priority because there are workarounds to the problem?

EisenbergEffect commented:
@davismj Didn't we recently have a fundamental fix to the router that removed the need for this?

davismj commented:
I don't think we need this. As far as I understand, any issues this would solve would be better solved by returning a promise for the async values in the activate callback. This would ensure that the values are properly defined at bind(). @CasiOo @michie1 is there anything I'm missing?

CasiOo commented:
I haven't had a personal need for any async href feature in any of our projects just yet, and I'm unsure if this is a frequently requested feature.

JoshMcCullough commented:
I [commented here](https://github.com/aurelia/templating-router/issues/21#issuecomment-350601920) why we load data in `attached` instead of `activate`. That's our use case; open to suggestions.

michie1 commented:
@davismj I don't mind if this pull request is getting rejected and I can't provide any use cases. Maybe it's better to close this PR and let the discussion continue in #21.
